### PR TITLE
feat: Initial Improvements

### PR DESCRIPTION
**Pull request by ianwestcott**
## Considerations

- Allow namespacing of Keycloak resources using an environment name
- Make various resources within the module optionally overridable by passing the name or ARN of an existing resource instead:
  - ACM certificate
  - ECR repo
  - ECS cluster
  - Database subnet group
  - S3 bucket for ELB logs
- Remove VPC and networking resources in favor of passing existing resources as vars
- Use Secrets Manager for database and admin passwords
- Use temporary, randomly-generated database password to avoid leaking password in state, and ignore password changes
- Scrub references to Integsoft and MBTA to ensure portability
- Add .gitignore, Terraform dependency lockfile, and README

### CODE REVIEW

File: loadbalancer.tf [hunk 1]
```diff
@@ -19,43 +24,41 @@ resource "aws_security_group" "load-balancer-sg" {
...
-  subnets            = aws_subnet.public.*.id
+  subnets            = var.public_subnets
   security_groups    = [aws_security_group.load-balancer-sg.id]
 
   # LB access logs
   access_logs {
-    bucket  = aws_s3_bucket.mbta-lb-access-logs.bucket
-    prefix  = "lb-keycloak"
+    bucket  = local.lb_log_bucket
+    prefix  = "keycloak-${var.environment}"
```
[the remark this answers is not in the document]
Comment (paulswartz) [reply loadbalancer.tf:42]: should the prefix be something like `keycloak-${var.environment}-alb` to distinguish them from application logs?

**Review by paulswartz — COMMENTED** [review]
general comments:
- I'm not sure the `Name` values are helpful
- The `Project` names should also include the name of the environment

File: vars.tf [hunk 2]
```diff
@@ -1,52 +1,80 @@
...
+  type        = string
+  description = "Keycloak database username"
 }
 
-variable "availability_zones" {
-  description = "List of availability zones"
+variable "ecr_keycloak_image_url" {
+  type        = string
+  description = "(optional) ECR Keycloak Image location. Defaults to creating a new ECR repo"
+  default     = null
```
Comment (grejdi-mbta) [inline vars.tf:31]: Where are we getting our Keycloak image from? Should we default to that.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# Keycloak Terraform Module
+
+A [Terraform](https://www.terraform.io/) [module](https://www.terraform.io/docs/language/modules/index.html) for creating and managing a [Keycloak](https://www.keycloak.org/) environment in AWS Elastic Container Service. 
+
+Co-authored by [Integsoft](https://www.integsoft.com/home.html) and MBTA.
```
[the remark this answers is not in the document]
Comment (ianwestcott) [reply README.md:5]: Should the link be to the Customer Technology microsite, or to MBTA.com?

File: loadbalancer.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (ianwestcott) [reply loadbalancer.tf:42]: I believe only S3 is supported as a destination for ELB access logs. It's possible to forward them to Splunk using bucket events and Lambda, which I tasked for future implementation. So in this case, the bucket would only exist for the purpose of storing ELB access logs, hence the naming convention.

File: vars.tf (hunk 2 above)
Comment (ianwestcott) [reply vars.tf:31]: The module uses the `null` value to determine whether to create a new ECR repo, so that has to be the default.

Comment (grejdi-mbta) [reply vars.tf:31]: @ianwestcott Where are we getting our ECS image though? Are we creating it ourselves, or downloading something from Docker Hub, such as [this](https://registry.hub.docker.com/r/jboss/keycloak#!)? I'm not too familiar with ECS structure, so sorry if I'm merging concepts here. If we are creating a plug and play module for others to use, then defaulting to the docker image url would make sense. And then the comparison would look something like this: `count = var.ecr_keycloak_image_url == ...local.keycloak_image_url.default... ? 1 : 0`

Comment (ianwestcott) [reply vars.tf:31]: @grejdi-mbta that's actually an open question that we're still working out with the vendor, but our hope is that we'll be building a Docker image from a Dockerfile in a separate MBTA repo. If there's a publicly-available fallback (e.g. the stock Keycloak image available in Docker Hub), I'd rather document that in the README as part of some "how to use this module" documentation, which I plan to add in the future. In either case, uploading the Docker image to ECR would happen outside of Terraform.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (paulswartz) [reply README.md:5]: I'd vote for Customer Technology.

**Comment by ianwestcott** [comment]
I opted to resolve the issues with tags by replacing all hard-coded tags with an optional var: https://github.com/mbta/terraform-keycloak-sso/pull/1/commits/261259e7bd275dea6f843b55d75bafc8ee2cc44d

File: certificate.tf [hunk 4]
```diff
@@ -1,15 +1,17 @@
-# TODO will be removed
+# certificate generation is optional, only used if var isn't passed
 resource "tls_private_key" "keycloak-pk" {
+  count     = var.acm_certificate_arn == null ? 1 : 0
```
[the remark this answers is not in the document]
Comment (ianwestcott) [reply certificate.tf]: How's this? https://github.com/mbta/terraform-keycloak-sso/pull/1/commits/c3529a6e8e0ed0e52688ef3a16662ab885e85339

File: iam.tf [hunk 5]
```diff
@@ -10,13 +10,13 @@ data "aws_iam_policy_document" "assume_role_policy" {
 }
 
 resource "aws_iam_role" "ecs-execution-task-role" {
```
[the remark this answers is not in the document]
Comment (ianwestcott) [reply iam.tf]: https://github.com/mbta/terraform-keycloak-sso/pull/1/commits/e03b2d2ca8736e082274e6486e0291d201abdaa4

File: keycloak.tf [hunk 6]
```diff
@@ -1,37 +1,43 @@
+locals {
```
[the remark this answers is not in the document]
Comment (ianwestcott) [reply keycloak.tf:1]: https://github.com/mbta/terraform-keycloak-sso/pull/1/commits/20e0641ae6761db39db78ddc49cd295fd6aefb72

**Comment by ianwestcott** [comment]
@grejdi-mbta for some reason GitHub isn't allowing me to request a new review from you, but kindly have another look when you have a chance? 🙏

**Review by grejdi-mbta — APPROVED** [review]

**Review by paulswartz — APPROVED** [review]
🍰

**Pull request merged**